sound/soc/sof/topology.c

```diff
@@ -952,10 +932,38 @@ static int sof_parse_tokens(struct snd_soc_component *scomp,
 		/* next array */
 		array = (struct snd_soc_tplg_vendor_array *)((u8 *)array
 			+ asize);
+
+		/*
+		 * offset needs to be incremented when a complete set of tokens
+		 * has been found.
+		 */
+		if (found == count) {
+			offset += object_size;
+			total += found;
+			found = 0;
+		}
```

Comment:
Thanks @juimonen ! I know @ranj063 @lyakh, this is less compact, but at least for me this is easier to understand.

Now the code simpler, maybe the comment could be just simply "/* move to the next set */" ...?

And more q: if the topology data is corrupt, are we sure nothing bad happens if "found > count" above. To me it looks ok, you end up parsing invalid data until priv_size goes to zero, but please check.

Reply:
ok, changed to doxygen style and simplified the other comment. I guess possible corruption would occur also with old version if ever. I f I read the code correctly, the word_tokens etc sub-functions loop "count" times for tokens, so found should never be > count... ? As you said basically we should not read past the memory area, but can't guarantee that some topologies could cause issues overall with our parsing...